test/modeling_library/product.jl

```diff
@@ -0,0 +1,89 @@
+discrete_product = ProductDistribution(bernoulli, binom)
+
+@testset "product of discrete distributions" begin
+    @test is_discrete(discrete_product)
+    grad_bools = (has_output_grad(discrete_product), has_argument_grads(discrete_product)...)
+    @test grad_bools == (false, true, false, true)
+
+    p1 = 0.5
+    (n, p2) = (3, 0.9)
+
+    # random
+    x = unrelated_pair_product(p1, n, p2)
+
+    # logpdf
+    x = (true, 2)
+    actual = logpdf(unrelated_pair_product, x, p1, n, p2)
+    expected = logpdf(bernoulli, x[1], p1) + logpdf(binom, x[2], n, p2)
+    @test isapprox(actual, expected)
+
+    # test logpdf_grad against finite differencing
+    f = (x, p, mu, std) -> logpdf(unrelated_pair_product, x, p, mu, std)
+    args = (x, p, mu, std)
+    actual = logpdf_grad(unrelated_pair_product, args...)
+    for (i, b) in enumerate(grad_bools)
+        if b
+            @test isapprox(actual[i], finite_diff(f, args, i, dx))
+        end
+    end
+end
+
+continuous_product = ProductDistribution(uniform, normal)
+
+@testset "product of continuous distributions" begin
+    @test !is_discrete(continuous_product)
+    grad_bools = (has_output_grad(continuous_product), has_argument_grads(continuous_product)...)
+    @test grad_bools == (true, true, true, true, true)
+
+    (low, high) = (-0.5, 0.5)
+    (mu, std) = (0.0, 1.0)
+
+    # random
+    x = continuous_product(low, high, mu, std)
```

Comment:
As above, we should add a type check and support check here (though we probably don't need a support check for the normally distributed variable).